src/components/button/index.tsx

```diff
@@ -1,18 +1,30 @@
 import React, { FC, MouseEvent, ReactNode } from 'react';
 import { Button as ChakraButton } from '@chakra-ui/react';
 
-interface Props {
+type SharedProps = {
   variant?: 'primary' | 'secondary';
   size?: 'md' | 'lg';
   isDisabled?: boolean;
-  onClick: (event: MouseEvent<HTMLElement>) => void;
   children: ReactNode;
-}
+};
+
+type ButtonProps = SharedProps & {
+  type?: 'button';
+  onClick: (event: MouseEvent<HTMLElement>) => void;
+};
+
+type SubmitProps = SharedProps & {
+  type?: 'submit';
+  onClick?: (event: MouseEvent<HTMLElement>) => void;
+};
+
+type Props = ButtonProps | SubmitProps;
 
 const Button: FC<Props> = ({
   variant = 'primary',
   size = 'lg',
   isDisabled = false,
+  type = 'button',
```

Comment:
can you confirm this behavior with the types:

When I do not specify a type (which defaults to "button") it does not complain if I do not pass an `onClick` handler. If I however specify the type explicitly to `button` it does complain.

Reply:
ran into the same issue, made a quick PR: [link the ticket here]

Reply:
That was intended. `onClick` was mandatory for `button` but we wanted to make it optional for `submit`

Reply:
What I mean is that since type=button is the default, the following should raise a TS error but it does not: `<Button>My button</Button>`. If I do this: `<Button type="button">My button</Button>` it does complain. But the two versions are identical

Reply:
Hmm, let me see what I can do about that, that wasn't intended. It could be that the `type` needs to be mandatory for this to work as expected

Reply:
I manage to achieve that with dependent types.
![Screenshot 2022-08-10 at 09 51 22](https://user-images.githubusercontent.com/144707/183845724-0b0fd54d-2803-4ed5-8383-68e207e79323.png)

The error we have when explicitly assigning `type` to `button` is less helpful due to how types need to be structured 
![Screenshot 2022-08-10 at 09 54 10](https://user-images.githubusercontent.com/144707/183846319-91679446-0af5-4a77-bf35-ba4e74bdb189.png)
(Basically, the `type` needs to be mandatory to switch the `onClick` enforcing on and off, otherwise `undefined` will match both clauses and we can't achieve it). 
It looks like a good compromise to me, but let me know what you think

Reply:
sounds good to me 👍